### DISCUSSION TITLE
Minimap similar to VSCode

### DISCUSSION THREAD

**Discussion by radian628** [DesModder/DesModder#845] [Ideas]
I think a feature like VSCode's minimap (that bar that shows a zoomed out view of the code and lets you click to quickly navigate) would be extremely helpful for organization and debugging with really large graphs. I envision that it wouldn't actually do any equation rendering (too slow), but it would show the following:

- Colored bars would represent displayed equations, points, inequalities, etc.
- Notes or folders would be indicated prominently to make navigation easier. There could even be some kind of string to show/hide them depending on what the notes are used for.
- Any errors would be highlighted (including highlighting a folder if they're within collapsed folders). Undefined variables could also be highlighted if an undefined variable is unexpected behavior.

**Comment by jared-hughes** [comment]
## Minimap

Here's VS Code's minimap

It shows the code (zoomed out heavily) and highlights:
- error locations
- currently-highlighted block
- cursor location
- diff from last commit

A collapsed block is collapsed in the minimap as well

![image](https://user-images.githubusercontent.com/20214911/137430126-7bb4dafe-6ba4-4efb-a0ea-b17365c9e606.png)

The DesModder version could show the equations (again zoomed) and highlight:
- error locations
  - this should respect the choices for hidden errors
  - errors are also marked at the folder position for collapsed folders
- currently-highlighted folder
- cursor location
- diff from last save would be a cool feature
- colors (and opacities?) of displayed items

## Outline

This is the google docs outline. I don't think it's wanted because it would be pretty similar to just collapsing all folders (see #236)

![image](https://user-images.githubusercontent.com/20214911/137429873-350453cf-6b7a-4177-bbf0-0200cc887f08.png)


## Other points

Could you give examples as to where undefined variables would be unexpected behavior?

**Comment by radian628** [comment]
> This is the google docs outline. I don't think it's wanted because it would be pretty similar to just collapsing all folders

I think it still might be useful to display the names of folders in a small font (or even just showing them when hovering), though I agree a full-blown outline like that of Google Docs would be redundant.

> Could you give examples as to where undefined variables would be unexpected behavior?

Here's two examples I was able to think of, though there are many more: The first is if you want to display some equation that contains an undefined variable. The undefined variable might prohibit it from being displayed. A second example is that of using undefined variable as a list bound, which causes an error. As of now, though such issues can be debugged, they can be difficult to trace in larger graphs, and thus showing which variables are undefined will likely speed up that debugging process. And while in theory any value could constitute unexpected behavior, I think undefined is often one of the biggest culprits.

**Comment by jared-hughes** [comment]
An initial attempt to this is https://github.com/DesModder/DesModder/pull/747, which used to be part of https://github.com/DesModder/DesModder/pull/743, providing the following demo image:

![image](https://github.com/user-attachments/assets/836b102e-a571-4f2d-95e5-45b71e2981ff)
